validator/src/journal/block_manager.rs

```diff
@@ -0,0 +1,1039 @@
+    }
+
+    fn get_block_by_block_id(&self, block_id: &str) -> Option<&Block> {
+        self.block_by_block_id
+            .get(block_id)
+            .map(|&(ref block, _, _)| block)
+    }
+
+    fn insert_block_by_block_id(&mut self, block: Block, internal: i64, external: i64) {
+        self.block_by_block_id
+            .insert(block.header_signature.clone(), (block, internal, external));
+    }
+
+    fn get_block_from_main_cache_or_blockstore_name(
+        &self,
+        block_id: &str,
+    ) -> (Option<&Block>, Option<&str>) {
+        let block = self.get_block_by_block_id(block_id);
+        if block.is_some() {
+            (block, None)
+        } else {
+            (None, self.anchors.get_blockstore_name(block_id))
+        }
+    }
+
+    fn get_block_from_blockstore<'a>(
+        &'a self,
+        block_id: String,
+        store_name: &str,
+    ) -> Result<Option<&'a Block>, BlockManagerError> {
+        Ok(self.blockstore_by_name
+            .get(store_name)
+            .ok_or(BlockManagerError::UnknownBlockStore)?
+            .get(vec![block_id])
+            .nth(0))
+    }
+
+    fn get_block_from_any_blockstore(&self, block_id: String) -> Option<&Block> {
+        self.blockstore_by_name
+            .values()
+            .find(|blockstore| blockstore.get(vec![block_id.clone()]).count() > 0)
+            .map(|blockstore| blockstore.get(vec![block_id]).nth(0).unwrap())
+    }
+
+    pub fn branch<'a>(&'a self, tip: &str) -> Box<Iterator<Item = &Block> + 'a> {
+        Box::new(BranchIterator::new(self, tip.into()))
+    }
+
+    pub fn branch_diff<'a>(
+        &'a self,
+        tip: &str,
+        exclude: &str,
+    ) -> Box<Iterator<Item = &Block> + 'a> {
+        Box::new(BranchDiffIterator::new(self, tip, exclude))
+    }
+
+    pub fn ref_block(&mut self, block_id: &str) -> Result<(), BlockManagerError> {
+        match self.block_by_block_id.get_mut(block_id) {
+            Some(&mut (_, _, ref mut external_ref_count)) => {
+                *external_ref_count += 1;
+                Ok(())
+            }
+            None => {
+                if self.anchors.contains(block_id) {
+                    self.anchors.increase_external_ref_count(block_id);
+                    Ok(())
+                } else {
+                    Err(BlockManagerError::UnknownBlock)
+                }
+            }
+        }
+    }
+
+    /// Starting at a tip block, if the tip block's ref-count drops to 0,
+    /// remove all blocks until a ref-count of 2 is found.
+    pub fn unref_block(&mut self, tip: &str) -> Result<(), BlockManagerError> {
+        let ref_count_block_id_result = self.lower_tip_blocks_refcount(tip);
+
+        let mut blocks_to_remove = vec![];
+
+        let mut optional_new_tip = None;
+
+        let (external_ref_count, internal_ref_count, block_id) = ref_count_block_id_result?;
+
+        if external_ref_count <= 0 && internal_ref_count <= 0 {
```

Comment:
I think this should be split up into checking whether the count is less than 0, which should be a panic, and checking if the count is 0, which should cause a purge.